File: tests/test_dimensionalisablefield.py
"""Tests for dimensionalisablefield.c (the scaled-field wrapper).

Driven through the C test executable tests/c/test_dimensionalisablefield,
which builds a five-node field with scaling 2.5, fills it with 1..5, and
exercises scale, unscale, duplicate, the local-vector factory, and the
scaling query. Contract clauses exercised: scaling multiplies by the
stored factor, unscaling inverts it exactly, the duplicate carries the
scaling and domain count, and the serial local vector matches the mesh
size. See docs/How-to/build_tests.md for the tier system.
"""

from __future__ import annotations

import pytest

pytestmark = [pytest.mark.unit, pytest.mark.timeout(30)]

# Values hard-coded in the C executable.
SCALING = 2.5
FIRST_VALUE = 1.0
MESH_SIZE = 5


def test_scale_unscale_round_trip_is_exact(c_test):
    """Scaling multiplies by the stored factor and unscaling inverts it.

    The field starts unscaled with first value 1.0; after Scale the
    stored entry is 2.5, and after Unscale it returns to 1.0 exactly
    (a multiply and divide by the same factor, no accumulation). The
    unit scaling factor is the edge the wrapper must not special-case:
    with 2.5 the scaled and raw values differ by 150 percent, so a
    no-op Scale cannot pass.
    """
    out = c_test('test_dimensionalisablefield')

    # rel=1e-15: one multiplication, roundoff only.
    assert out['scaled0'] == pytest.approx(SCALING * FIRST_VALUE, rel=1e-15)
    # No-op guard: scaling must actually change the stored values.
    assert abs(out['scaled0'] - out['original0']) > 1.0
    # Round trip: unscale exactly inverts scale.
    assert out['roundtrip0'] == pytest.approx(out['original0'], rel=1e-15)
    assert out['original0'] == pytest.approx(FIRST_VALUE, rel=1e-15)


def test_duplicate_carries_scaling_and_layout(c_test):
    """A duplicated field reports the parent's scaling and domain count.

    The duplicate is created over the same serial DMDA, so it has one
    domain, the parent's 2.5 scaling, and a local vector of the mesh
    size (the limit input: a single-domain, single-rank layout with no
    ghost points).
    """
    out = c_test('test_dimensionalisablefield')

    assert out['dup_scaling0'] == pytest.approx(SCALING, rel=1e-15)
    # Scaling discrimination: the duplicate must carry 2.5, not the
    # unit scaling a fresh field would default to.
    assert abs(out['dup_scaling0'] - 1.0) > 1.0
    assert int(out['num_domains']) == 1
    assert int(out['nlocal']) == MESH_SIZE

File: tests/c/test_dimensionalisablefield.c
static const char help[] =
    "Exercises the DimensionalisableField wrapper on a small DMDA:\n"
    "create, fill, scale, unscale, duplicate, local vector, and the\n"
    "scaling query. Prints one JSON object to stdout; the pytest\n"
    "wrapper owns all assertions.\n";

#include <petsc.h>

#include "dimensionalisablefield.h"

#define TEST_N 5
#define TEST_SCALING 2.5

int main(int argc, char **argv)
{
  PetscErrorCode         ierr;
  DM                     dm;
  DimensionalisableField f, fdup;
  Vec                    v, vlocal;
  PetscScalar            scalings[1] = {TEST_SCALING};
  PetscScalar            dupScaling[1];
  const PetscScalar      *arr;
  PetscScalar            original0, scaled0, rescaled0, roundtrip0;
  PetscInt               i, numDomains, nlocal;

  ierr = PetscInitialize(&argc, &argv, NULL, help);
  if (ierr) return (int)ierr;

  ierr = DMDACreate1d(PETSC_COMM_WORLD, DM_BOUNDARY_NONE, TEST_N, 1, 1, NULL, &dm);CHKERRQ(ierr);
  ierr = DMSetUp(dm);CHKERRQ(ierr);

  ierr = DimensionalisableFieldCreate(&f, dm, scalings, PETSC_FALSE);CHKERRQ(ierr);
  ierr = DimensionalisableFieldSetName(f, "test field");CHKERRQ(ierr);
  ierr = DimensionalisableFieldSetUnits(f, "test units");CHKERRQ(ierr);

  /* fill the global vector with 1..N so index slips are visible */
  ierr = DimensionalisableFieldGetGlobalVec(f, &v);CHKERRQ(ierr);
  for (i = 0; i < TEST_N; ++i) {
    ierr = VecSetValue(v, i, (PetscScalar)(i + 1), INSERT_VALUES);CHKERRQ(ierr);
  }
  ierr = VecAssemblyBegin(v);CHKERRQ(ierr);
  ierr = VecAssemblyEnd(v);CHKERRQ(ierr);

  ierr = VecGetArrayRead(v, &arr);CHKERRQ(ierr);
  original0 = arr[0];
  ierr = VecRestoreArrayRead(v, &arr);CHKERRQ(ierr);

  /* scale multiplies by the stored scaling, unscale inverts it */
  ierr = DimensionalisableFieldScale(f);CHKERRQ(ierr);
  ierr = VecGetArrayRead(v, &arr);CHKERRQ(ierr);
  scaled0 = arr[0];
  ierr = VecRestoreArrayRead(v, &arr);CHKERRQ(ierr);

  /* scaling an already-scaled field is the documented no-op guard:
     it warns and must leave the values untouched */
  ierr = DimensionalisableFieldScale(f);CHKERRQ(ierr);
  ierr = VecGetArrayRead(v, &arr);CHKERRQ(ierr);
  rescaled0 = arr[0];
  ierr = VecRestoreArrayRead(v, &arr);CHKERRQ(ierr);

  ierr = DimensionalisableFieldUnscale(f);CHKERRQ(ierr);
  ierr = VecGetArrayRead(v, &arr);CHKERRQ(ierr);
  roundtrip0 = arr[0];
  ierr = VecRestoreArrayRead(v, &arr);CHKERRQ(ierr);

  /* the duplicate carries the scaling and domain layout */
  ierr = DimensionalisableFieldDuplicate(f, &fdup);CHKERRQ(ierr);
  numDomains = 1;
  ierr = DimensionalisableFieldGetScaling(fdup, &numDomains, dupScaling);CHKERRQ(ierr);

  /* the local vector matches the serial DMDA layout */
  ierr = DimensionalisableFieldCreateLocalVec(f, &vlocal);CHKERRQ(ierr);
  ierr = VecGetLocalSize(vlocal, &nlocal);CHKERRQ(ierr);

  ierr = PetscPrintf(PETSC_COMM_WORLD, "{\n");CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"original0\": %.17g,\n", (double)original0);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"scaled0\": %.17g,\n", (double)scaled0);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"rescaled0\": %.17g,\n", (double)rescaled0);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"roundtrip0\": %.17g,\n", (double)roundtrip0);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"scaling\": %.17g,\n", (double)TEST_SCALING);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"dup_scaling0\": %.17g,\n", (double)dupScaling[0]);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"num_domains\": %d,\n", (int)numDomains);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "\"nlocal\": %d\n", (int)nlocal);CHKERRQ(ierr);
  ierr = PetscPrintf(PETSC_COMM_WORLD, "}\n");CHKERRQ(ierr);

  ierr = VecDestroy(&vlocal);CHKERRQ(ierr);
  ierr = DimensionalisableFieldDestroy(&fdup);CHKERRQ(ierr);
  ierr = DimensionalisableFieldDestroy(&f);CHKERRQ(ierr);
  ierr = DMDestroy(&dm);CHKERRQ(ierr);
  ierr = PetscFinalize();
  return 0;
}
